### ISSUE TITLE
Take picture

### ISSUE THREAD

**Issue by pyrog** [Zverik/every_door#184]
1. Simply for documenting a POI
2. Also to publish it on the web (Mapillary, Wikimedia Commons…) and set the appropriate tags (mapillary, wikimedia_commons, image…)

In the first case, pictures could be stored in the phone gallery according user preferences (SD card, phone memory).

**Comment by Binnette** [comment]
Note that [StreetComplete](https://github.com/streetcomplete/StreetComplete) allows users to add a picture on a note. Ex:
- https://www.openstreetmap.org/note/3176448
- https://westnordost.de/p/99124.jpg

May be @Zverik can host a similar solution on https://ilya.zverev.info ?

Server side @westnordost uses a PHP application: https://github.com/streetcomplete/sc-photo-service
And client side, he implements the upload in this class: [StreetCompleteImageUploader](https://github.com/streetcomplete/StreetComplete/blob/master/app/src/main/java/de/westnordost/streetcomplete/data/osmnotes/StreetCompleteImageUploader.kt)

Have a nice day

**Comment by JLZIMMERMANN** [comment]
The problem is that the pictures taken in Streetcomplete are photo-notes and are note taken to be linked with the OSM objetct.

**Comment by pyrog** [comment]
The problems with StreetComplete are:
- pictures are reduced (so it's difficult or impossible to read something i.e. a house number)
- pictures are automatically deleted after a while when notes are closed.

**Comment by pyrog** [comment]
When I collaborate with Robert Koch to OsmHydrant few years ago, previous Flickr OsmHydrant pictures were uploaded to Wikimedia Commons, and since, the web app upload new pictures directly to Commons.

But anyway, the ideal app should be agnostic : take a picture and publish it to the preferred repository (Flickr, Wikimedia Commons, Mapillary, a GitHub repo, a "Géo Commun"…).

If applicable, metadata in the repo could be set (i.e. `date`, `lat`, `lon`, ±`osm id`, ±`wikidata item`…) and an osm tag too (`mapillary`, `wikimedia_commons`, `image`)

**Comment by JLZIMMERMANN** [comment]
Have you checked Commons (Android) app ? It gives option to push pictures in wikimedia servers. In my idea if we are able to create OSM categories we would upload located pictures that would be connected with OSM pictures like OSMhydrant did at the end.

**Comment by j9d3it** [comment]
[Imgur has an anonymous image upload API.](https://apidocs.imgur.com/#c85c9dfc-7487-4de2-9ecd-66f727cf3139) This could work as well.

**Comment by j9d3it** [comment]
> [Imgur has an anonymous image upload API.](https://apidocs.imgur.com/#c85c9dfc-7487-4de2-9ecd-66f727cf3139) This could work as well.

Update!!

[Imgur will remove anonymously-uploaded images that have not been accessed.](https://help.imgur.com/hc/en-us/articles/14415587638029-Imgur-Terms-of-Service-Update-April-19-2023-)

**Comment by Zverik** [comment]
So I'm warming up to the idea of taking photos with the app for notes. The latest version already has two types of notes and a dedicated server, so no reason not to add a third one, for photos. They can be stored for half a year with an option of exporting them in bulk to re-upload to e.g. wikimedia commons.

**Comment by tordans** [comment]
I see the appeal to add a custom image feature to every door. I still have hopes that at some point we will build a solution that works for OSM general or at least for alle editors(*). Until then, a few more links

- https://github.com/pietervdvn/MapComplete/issues/1451 holds information on how MapComplete plant so handle user image uploads in the future. Those images are attached to the OSM object, though, not OSM Notes.
- https://github.com/streetcomplete/StreetComplete/issues/5489#issuecomment-1962339611 is a good summary on the status of this feature in StreetComplete and has a few links on what was tried on the OSM website / operations side of things.

---

(*) There was [some activity on micro grants](https://github.com/osmfoundation/ewg_bidding/issues/14) a short while ago, so maybe those will come again and maybe this is something to build in such a project?

**Comment by tordans** [comment]
Update on …

> * [Upload pictures to Panoramax/Geovisio instead of Imgur pietervdvn/MapComplete#1451](https://github.com/pietervdvn/MapComplete/issues/1451) holds information on how MapComplete plant so handle user image uploads in the future. Those images are attached to the OSM object, though, not OSM Notes.

https://github.com/openstreetmap/iD/issues/9941#issuecomment-2037888222 suggests there will be some progress on the Panoramax side of things if the funding does happen.

**Comment by Coehill** [comment]
Mapcomplete now has Panoramax integration
https://github.com/pietervdvn/MapComplete/issues/1451#issuecomment-2380550206

**Comment by tordans** [comment]
In https://github.com/bryceco/GoMap/issues/822 I asked how this MapComplete feature could be shared with GoMap but it [sounds like](https://github.com/bryceco/GoMap/issues/822#issuecomment-2433339208) this is not planned.

**Comment by Zverik** [comment]
Linking StreetComplete ticket just in case: https://github.com/streetcomplete/StreetComplete/issues/5939

**Comment by westnordost** [comment]
So, in order to link photos with OSM elements generally, this definitely would need to be an overlay in which you can upload pictures for elements, because I don't think this would work as a quest.

Even for an overlay, it's not quite clear for which elements it should be possible to add pictures. Only places and things? More? Regardless, that overlay would get quite busy. Displaying **which** elements already have a picture linked would probably be possible, but to display these pictures (right on the map) quite impossible. Even displaying them when opening the form is a problem for an app that should otherwise work fully offline.

It would also be quite an oddball amongst the overlays, as you are not directly contributing missing data, rather, metadata or something that might be useful later. For the user, it would be quite unclear which POIs should have a picture and which shouldn't. Also, who is **only** going to add pictures of elements? (I mean, that's the purpose of overlays in StreetComplete, to focus on **one** thing.)

So for this reason, to be honest, I think this whole idea of adding pictures to an element is a feature that would be more fitting for a general-purpose (POI) editor like iD, EveryDoor or Vespucci. E.g. in the view where all the possible (supported) feature properties are displayed, have an additional `image` field that has a [📷] button with which one can easily add a foto. That foto is in parallel uploaded to a persistent image hosting service and the URL to it is added in the `image` tag of the element.

(For now,) I will close this as a will not fix. I think this is a bit out of scope for StreetComplete. And I am happy that the EveryDoor developer is thinking about implementing such a thing.


But which service to use for that? Really Panoramax?

## Panoramax

Pros:
- Reporting mechanism: Hides pictures when reported
- automatically license plates and faces are blurred (might backfire, though)

Cons:
- that service is not really made for taking pictures of things somewhere, it's seems a bit of a mis-use or unintended usage of Panoramax to me to use it essentially as a persistent photo-hosting service.
- user needs to authorize with OSM twice (first for the editor app, second for picture upload)

I don't know, maybe a Panoramax instance which has all those `image`s from the OSM elements would indeed open up new use cases and in general be a convenient way to access these. But then, that Panoramax instance should show **all** the pictures linked from the `image` tag in OSM elements, no?

## Alternative

To avoid the cons of Panoramax, one might also just host a simple image hosting service akin to https://github.com/streetcomplete/sc-photo-service but with a different cleanup-mechanism: The cleanup doesn't look at whether the note in which the pictures were mentioned is still open, but looks if the image is still linked from the associated element in OSM. 

Pros:
- no need to authorize twice. Same "photo activation" mechanism is used as for the StreetComplete photo notes
- very lightweight (just a bunch of PHP scripts, a MariaDB database and lots of webspace)
- straightforward moderation/reporting mechanism: When the link is severed, the image will be deleted (by a cronjob, i.e. after some time)

Cons:
- license plates and faces are not blurred automatically. (It would need to be implemented)

sc-photo-upload could be forked, or even extended, to be able to do both.
The limit in picture size for image note uploads, by the way, is something that is enforced client-side by StreetComplete

**Comment by westnordost** [comment]
(oops, posted in the wrong issue)

**Comment by Zverik** [comment]
regarding the single sign-on: maybe it's possible to extend [Panoramax authentication](https://gitlab.com/panoramax/server/api/-/blob/develop/geovisio/utils/auth.py?ref_type=heads) to pass an OSM token with a client id, it would validate the user name and id, and return a token with which to upload photos.

**Comment by Zverik** [comment]
So [I did it lol](https://gitlab.com/zverik/panoramax-api/-/commit/99d8978c5db5755da9b57f1c3bbbbdf7db7e96fd): a new endpoint that uses a stolen `access_token` to just get the info. Not sure if good, but I successfully registered on my local Panoramax instance using `cli-oauth2` python library and this endpoint.

**Comment by JLZIMMERMANN** [comment]
> In [bryceco/GoMap#822](https://github.com/bryceco/GoMap/issues/822) I asked how this MapComplete feature could be shared with GoMap but it [sounds like](https://github.com/bryceco/GoMap/issues/822#issuecomment-2433339208) this is not planned.

Now it's updated as working

![Image](https://github.com/user-attachments/assets/ad314ad9-6b4e-4768-b2e1-62a523fa4da6)
